exonum/src/node/consensus.rs

```diff
@@ -546,24 +595,42 @@ impl NodeHandler {
 
     /// Checks if the transaction is new and adds it to the pool. This may trigger an expedited
     /// `Propose` timeout on this node if transaction count in the pool goes over the threshold.
-    pub fn handle_tx(&mut self, msg: Verified<AnyTx>) -> Result<(), failure::Error> {
+    ///
+    /// Before adding a transaction into pool, this method calls `BlockchainMut::check_tx` to
+    /// ensure that transaction passes at least basic checks. If `BlockchainMut::check_tx` fails,
+    /// transaction will be considered invalid and not stored to the pool (instead, its hash will
+    /// be stored in the temporary invalid messages set, so we will be able to detect a block/propose
+    /// with an invalid tx later).
+    ///
+    /// # Panics
+    ///
+    /// This function panics if it receives an invalid transaction for an already committed block.
+    pub(crate) fn handle_tx(&mut self, msg: Verified<AnyTx>) -> Result<(), failure::Error> {
         let hash = msg.object_hash();
 
         let snapshot = self.blockchain.snapshot();
         let schema = Schema::new(&snapshot);
-
         if contains_transaction(&hash, &schema.transactions(), self.state.tx_cache()) {
             bail!("Received already processed transaction, hash {:?}", hash)
         }
 
-        // TODO We have to check transaction correctness.
+        if let Err(e) = self.blockchain.check_tx(&msg) {
+            // Store transaction as invalid to know it if it'll be included into a proposal.
```

Comment:
It could make sense to add in the comment that marking a transaction as invalid **must** occur before calling `check_incomplete_proposes`, since the latter uses `invalid_txs` to recalculate the validity of proposals.